nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/db/event/TableInfo.java

```diff
@@ -0,0 +1,146 @@
+
+        TableInfo tableInfo = (TableInfo) o;
+
+        if (!databaseName.equals(tableInfo.databaseName)) return false;
+        if (!tableName.equals(tableInfo.tableName)) return false;
+        if (!tableId.equals(tableInfo.tableId)) return false;
+        return columns != null ? columns.equals(tableInfo.columns) : tableInfo.columns == null;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = databaseName.hashCode();
+        result = 31 * result + tableName.hashCode();
+        result = 31 * result + tableId.hashCode();
+        result = 31 * result + (columns != null ? columns.hashCode() : 0);
+        return result;
+    }
+
+    public static class Serializer implements org.apache.nifi.distributed.cache.client.Serializer<TableInfo> {
+
+        @Override
+        public void serialize(TableInfo value, OutputStream output) throws SerializationException, IOException {
+            StringBuilder sb = new StringBuilder(value.getDatabaseName());
+            sb.append(DB_TABLE_NAME_DELIMITER);
+            sb.append(value.getTableName());
+            sb.append(DB_TABLE_NAME_DELIMITER);
+            sb.append(value.getTableId());
+            List<ColumnDefinition> columnDefinitions = value.getColumns();
+            if (columnDefinitions != null && !columnDefinitions.isEmpty()) {
+                sb.append(DB_TABLE_NAME_DELIMITER);
+                sb.append(columnDefinitions.stream().map((col) -> col.getName() + DB_TABLE_NAME_DELIMITER + col.getType()).collect(Collectors.joining(DB_TABLE_NAME_DELIMITER)));
+            }
+            output.write(sb.toString().getBytes());
+        }
+    }
+
+    public static class Deserializer implements org.apache.nifi.distributed.cache.client.Deserializer<TableInfo> {
+
+        @Override
+        public TableInfo deserialize(byte[] input) throws DeserializationException, IOException {
+            // Don't bother deserializing if empty, just return null. This usually happens when the key is not found in the cache
+            if (input == null || input.length == 0) {
+                return null;
+            }
+            String inputString = new String(input);
+            String[] tokens = inputString.split(DB_TABLE_NAME_DELIMITER);
+            int numTokens = tokens.length;
+            if (numTokens < 3) {
+                throw new IOException("Could not deserialize TableInfo from the following value: " + inputString);
+            }
+            String dbName = tokens[0];
+            String tableName = tokens[1];
+            Long tableId;
+            try {
+                tableId = Long.parseLong(tokens[2]);
+            } catch (NumberFormatException nfe) {
+                throw new IOException("Illegal table ID: " + tokens[2]);
+            }
+            // Parse column names and types
+            List<ColumnDefinition> columnDefinitions = new ArrayList<>();
+            for (int i = 0; i < numTokens - 3; i += 2) {
```

Comment:
Hopefully only the provided Serializer is writing the values, but you're right that we should code defensively. I will check to ensure the entries exist before trying to create a ColumnDefinition, and throw an IOException if something's missing.